File: Hippo/Views/WebtoonCell.m
//
//  WebtoonCell.m
//  Hippo
//
//  Created by 전수열 on 13. 9. 1..
//  Copyright (c) 2013년 Joyfl. All rights reserved.
//

#import "WebtoonCell.h"

@implementation WebtoonCell

- (id)initWithReuseIdentifier:(NSString *)reuseIdentifier
{
	self = [super init];
	
	self.thumbnailView = [[UIImageView alloc] initWithFrame:CGRectMake( 0, 0, 55, 60 )];
	[self.contentView addSubview:self.thumbnailView];
	
	self.titleLabel = [[UILabel alloc] initWithPosition:CGPointMake( 65, 4 ) maxWidth:190];
	[self.contentView addSubview:self.titleLabel];
	
	self.countLabel = [[UILabel alloc] init];
	self.countLabel.backgroundColor = [UIColor orangeColor];
	self.countLabel.textColor = [UIColor whiteColor];
	self.countLabel.font = [UIFont boldSystemFontOfSize:11];
	self.countLabel.layer.cornerRadius = 6;
	[self.contentView addSubview:self.countLabel];
	
	self.artistLabel = [[UILabel alloc] initWithPosition:CGPointMake( 65, 24 ) maxWidth:190];
	self.artistLabel.font = [UIFont systemFontOfSize:12];
	self.artistLabel.textColor = [UIColor grayColor];
	[self.contentView addSubview:self.artistLabel];
	
	self.portalIconView = [[UIImageView alloc] initWithFrame:CGRectMake( 65, 41, 12, 12 )];
	[self.contentView addSubview:self.portalIconView];
	
	self.weekdayLabel = [[UILabel alloc] initWithPosition:CGPointMake( 81, 41 ) maxWidth:170];
	self.weekdayLabel.font = [UIFont systemFontOfSize:11];
	self.weekdayLabel.textColor = [UIColor grayColor];
	[self.contentView addSubview:self.weekdayLabel];
	
	self.subscribeButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
	self.accessoryView = self.subscribeButton;
	
	return self;
}

- (void)setWebtoon:(Webtoon *)webtoon
{
	_webtoon = webtoon;
	[self layoutContentView];
}

- (void)layoutContentView
{
	[self.thumbnailView setImageWithURL:[NSURL URLWithString:self.webtoon.thumbnail_url] placeholderImage:nil];
	
	self.titleLabel.text = self.webtoon.title;
	[self.titleLabel sizeToFit];
	
#warning self.webtoon.new_count로 접근하면 -[CFNumber retain]: message sent to deallocated instance 에러 발생.
	NSInteger newCount = [[self.webtoon valueForKey:@"new_count"] integerValue];
	if( newCount )
	{
		self.countLabel.text = [NSString stringWithFormat:@" %d ", newCount];
		[self.countLabel sizeToFit];
		self.countLabel.position = CGPointMake( self.titleLabel.frame.origin.x + self.titleLabel.frame.size.width + 5, self.titleLabel.frame.origin.y + 3 );
	}
	
	if( [self.webtoon.artists count] )
	{
		NSMutableArray *artists = [NSMutableArray array];
		for( NSDictionary *artist in self.webtoon.artists ) {
			[artists addObject:[artist objectForKey:@"name"]];
		}
		self.artistLabel.text = [artists componentsJoinedByString:@", "];
		[self.artistLabel sizeToFit];
	}
	
	self.portalIconView.image = [UIImage imageNamed:[NSString stringWithFormat:@"icon_%@.png", self.webtoon.portal]];
	
	if( self.webtoon.finished.boolValue )
	{
		self.weekdayLabel.text = @"완결";
	}
	else
	{
		NSMutableArray *weekdays = [NSMutableArray array];
		for( NSInteger i = 1; i < 7; i++ )
		{
			NSString *weekday = HippoWeekdays[i];
			if( [[self.webtoon valueForKey:weekday] boolValue] ) {
				[weekdays addObject:NSLocalizedString( weekday.uppercaseString, nil )];
			}
		}
		self.weekdayLabel.text = [weekdays componentsJoinedByString:@", "];
	}
	[self.weekdayLabel sizeToFit];
	
	[self.subscribeButton setTitle:self.webtoon.subscribed.boolValue ? @"구독중" : @"구독하기" forState:UIControlStateNormal];
	[self.subscribeButton sizeToFit];
}

@end

File: Hippo/Views/WebtoonCell.swift
//
//  WebtoonCell.swift
//  Hippo
//
//  Created by 전수열 on 10/3/14.
//  Copyright (c) 2014 Joyfl. All rights reserved.
//

import Foundation

class WebtoonCell: UITableViewCell {

    private var _webtoon: Webtoon?
    var webtoon: Webtoon? {
        get {
            return _webtoon?
        }
        set {
            self._webtoon = newValue
            let placeholder = UIImage(color: UIColor.whiteColor())
            if newValue!.picture? == nil {
                self.thumbnailView.image = placeholder
            } else {
                let url = NSURL(string: newValue!.picture!.url)
                self.thumbnailView.setImageWithURL(url, placeholderImage: placeholder)
            }
            self.titleLabel.text = newValue!.title
            self.artistLabel.text = newValue!.artistText()
            self.portalIconView.image = UIImage(named: "icon_\(newValue!.portal)")
            self.weekdayLabel.text = newValue!.weekdayText()

            let normalText = !newValue!.concluded ? __("Subscribe") : __("Collect")
            let selectedText = !newValue!.concluded ? __("Subscribing") : __("Collected")
            self.subscribeButton.setTitle(normalText, forState: UIControlState.Normal)
            self.subscribeButton.setTitle(selectedText, forState: UIControlState.Selected)
            self.subscribeButton.selected = newValue!.subscribing
        }
    }

    var thumbnailView = UIImageView()
    var titleLabel = UILabel()
    var countLabel = UILabel()
    var artistLabel = UILabel()
    var subscribeButton = SSBouncyButton()
    var portalIconView = UIImageView()
    var weekdayLabel = UILabel()


    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        self.contentView.addSubview(self.thumbnailView)
        self.contentView.addSubview(self.titleLabel)
        self.contentView.addSubview(self.countLabel)
        self.contentView.addSubview(self.artistLabel)
        self.contentView.addSubview(self.portalIconView)
        self.contentView.addSubview(self.weekdayLabel)
        self.contentView.addSubview(self.subscribeButton)

        self.thumbnailView.snp_makeConstraints { make in
            make.width.equalTo(55)
            make.height.equalTo(60)
        }

        self.titleLabel.snp_makeConstraints { make in
            make.left.equalTo(65)
            make.right.equalTo(self.subscribeButton.snp_left).with.offset(-5)
            make.top.equalTo(4)
        }

        self.countLabel.backgroundColor = UIColor.orangeColor()
        self.countLabel.textColor = UIColor.whiteColor()
        self.countLabel.font = UIFont.boldSystemFontOfSize(11)
        self.countLabel.layer.cornerRadius = 6
        self.countLabel.snp_makeConstraints { make in
            make.left.equalTo(self.titleLabel).with.offset(5)
            make.top.equalTo(self.titleLabel).with.offset(3)
        }

        self.artistLabel.font = UIFont.boldSystemFontOfSize(12)
        self.artistLabel.textColor = UIColor.grayColor()
        self.artistLabel.snp_makeConstraints { make in
            make.left.equalTo(65)
            make.right.equalTo(self.titleLabel)
            make.top.equalTo(24)
        }

        self.portalIconView.snp_makeConstraints { make in
            make.left.equalTo(65)
            make.top.equalTo(41)
            make.size.equalTo(CGSizeMake(12, 12))
        }

        self.weekdayLabel.font = UIFont.systemFontOfSize(11)
        self.weekdayLabel.textColor = UIColor.grayColor()
        self.weekdayLabel.snp_makeConstraints { make in
            make.left.equalTo(81)
            make.top.equalTo(41)
            make.width.lessThanOrEqualTo(170)
        }

        self.subscribeButton.titleLabel?.textAlignment = NSTextAlignment.Right
        self.subscribeButton.addTarget(self, action: "subscribeButtonDidPress",
            forControlEvents: UIControlEvents.TouchUpInside)
        self.subscribeButton.snp_makeConstraints { make in
            make.right.equalTo(self.contentView).with.offset(-10)
            make.centerY.equalTo(self.contentView)
            make.width.equalTo(60)
            make.height.equalTo(34)
        }
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    class var height: CGFloat {
        return 60;
    }

    func subscribeButtonDidPress() {
        self.subscribeButton.selected = !self.subscribeButton.selected

        var routeName = "subscribe_webtoon"
        if !self.subscribeButton.selected {
            routeName = "un" + routeName
        }

        Request.sendToRoute(routeName, parameters: ["webtoon_id": self.webtoon!.id],
            success: { (operation, responseObject) -> Void in
                RLMRealm.defaultRealm().beginWriteTransaction()
                self.webtoon!.subscribing = self.subscribeButton.selected
                RLMRealm.defaultRealm().commitWriteTransaction()
            },
            failure: { (operation, error) -> Void in
                println("\(operation.responseObject)")
                self.subscribeButton.selected = !self.subscribeButton.selected
            }
        )
    }
}
